// office-addin-usage-data/lib/usageData.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OfficeAddinUsageData = exports.ExpectedError = exports.UsageDataLevel = exports.UsageDataReportingMethod = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
/**
 * Specifies the usage data infrastructure the user wishes to use
 * @enum Application Insights: Microsoft Azure service used to collect and query through data
 * @deprecated Usage data reporting has been removed
 */
var UsageDataReportingMethod;
(function (UsageDataReportingMethod) {
    UsageDataReportingMethod["applicationInsights"] = "applicationInsights";
})(UsageDataReportingMethod = exports.UsageDataReportingMethod || (exports.UsageDataReportingMethod = {}));
/**
 * Level controlling what type of usage data is being sent
 * @enum off: off level of usage data, sends no usage data
 * @enum on: on level of usage data, sends errors and events
 * @deprecated Usage data reporting has been removed
 */
var UsageDataLevel;
(function (UsageDataLevel) {
    UsageDataLevel["off"] = "off";
    UsageDataLevel["on"] = "on";
})(UsageDataLevel = exports.UsageDataLevel || (exports.UsageDataLevel = {}));
/**
 * Defines an error that is expected to happen given some situation
 * @member message Message to be logged in the error
 */
class ExpectedError extends Error {
    constructor(message) {
        super(message);
        // need to adjust the prototype after super()
        // See https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
        Object.setPrototypeOf(this, ExpectedError.prototype);
    }
}
exports.ExpectedError = ExpectedError;
/**
 * Usage data class - all methods are now no-ops since telemetry has been removed
 * @deprecated Usage data reporting has been removed. This class is retained for API compatibility only.
 */
class OfficeAddinUsageData {
    constructor(usageDataOptions) {
        this.options = Object.assign({ usageDataLevel: UsageDataLevel.off }, usageDataOptions);
    }
    /** @deprecated No-op - usage data reporting has been removed */
    reportEvent(_eventName, _data) {
        return __awaiter(this, void 0, void 0, function* () { });
    }
    /** @deprecated No-op - usage data reporting has been removed */
    reportEventApplicationInsights(_eventName, _data) {
        return __awaiter(this, void 0, void 0, function* () { });
    }
    /** @deprecated No-op - usage data reporting has been removed */
    reportError(_errorName, _err) {
        return __awaiter(this, void 0, void 0, function* () { });
    }
    /** @deprecated No-op - usage data reporting has been removed */
    reportErrorApplicationInsights(_errorName, _err) {
        return __awaiter(this, void 0, void 0, function* () { });
    }
    /** @deprecated No-op - usage data reporting has been removed */
    usageDataOptIn(_testData, _testResponse) { }
    /** @deprecated No-op - usage data reporting has been removed */
    setUsageDataOff() { }
    /** @deprecated No-op - usage data reporting has been removed */
    setUsageDataOn() { }
    /** @deprecated Always returns false - usage data reporting has been removed */
    isUsageDataOn() {
        return false;
    }
    /** @deprecated Returns empty string - usage data reporting has been removed */
    getUsageDataKey() {
        return "";
    }
    /** @deprecated Always returns 0 - usage data reporting has been removed */
    getEventsSent() {
        return 0;
    }
    /** @deprecated Always returns 0 - usage data reporting has been removed */
    getExceptionsSent() {
        return 0;
    }
    /** @deprecated Always returns "off" - usage data reporting has been removed */
    getUsageDataLevel() {
        return UsageDataLevel.off;
    }
    /** @deprecated Returns error unchanged - usage data reporting has been removed */
    maskFilePaths(err) {
        return err;
    }
    /** @deprecated No-op - usage data reporting has been removed */
    reportException(_method, _err, _data) { }
    /** @deprecated No-op - usage data reporting has been removed */
    reportExpectedException(_method, _err, _data) { }
    /** @deprecated No-op - usage data reporting has been removed */
    reportSuccess(_method, _data) { }
    /** @deprecated No-op - usage data reporting has been removed */
    sendUsageDataException(_method, _err, _data) { }
    /** @deprecated No-op - usage data reporting has been removed */
    sendUsageDataSuccessEvent(_method, _data) { }
    /** @deprecated No-op - usage data reporting has been removed */
    sendUsageDataSuccessfulFailEvent(_method, _data) { }
    /** @deprecated No-op - usage data reporting has been removed */
    sendUsageDataEvent(_data) { }
}
exports.OfficeAddinUsageData = OfficeAddinUsageData;
