// office-addin-usage-data/lib/usageDataSettings.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.readUsageDataSettings = exports.groupNameExists = exports.writeUsageDataJsonData = exports.readUsageDataObjectProperty = exports.readUsageDataLevel = exports.readUsageDataJsonData = exports.readDeviceID = exports.modifyUsageDataJsonData = exports.needToPromptForUsageData = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
const usageData_1 = require("./usageData");
/**
 * @deprecated Usage data reporting has been removed - always returns false
 */
function needToPromptForUsageData(_groupName) {
    return false;
}
exports.needToPromptForUsageData = needToPromptForUsageData;
/**
 * @deprecated Usage data reporting has been removed - no-op
 */
function modifyUsageDataJsonData(_groupName, _property, _value) { }
exports.modifyUsageDataJsonData = modifyUsageDataJsonData;
/**
 * @deprecated Usage data reporting has been removed - returns empty string
 */
function readDeviceID() {
    return "";
}
exports.readDeviceID = readDeviceID;
/**
 * @deprecated Usage data reporting has been removed - returns undefined
 */
function readUsageDataJsonData() {
    return undefined;
}
exports.readUsageDataJsonData = readUsageDataJsonData;
/**
 * @deprecated Usage data reporting has been removed - always returns off
 */
function readUsageDataLevel(_groupName) {
    return usageData_1.UsageDataLevel.off;
}
exports.readUsageDataLevel = readUsageDataLevel;
/**
 * @deprecated Usage data reporting has been removed - returns undefined
 */
function readUsageDataObjectProperty(_groupName, _propertyName) {
    return undefined;
}
exports.readUsageDataObjectProperty = readUsageDataObjectProperty;
/**
 * @deprecated Usage data reporting has been removed - no-op
 */
function writeUsageDataJsonData(_groupName, _level) { }
exports.writeUsageDataJsonData = writeUsageDataJsonData;
/**
 * @deprecated Usage data reporting has been removed - always returns false
 */
function groupNameExists(_groupName) {
    return false;
}
exports.groupNameExists = groupNameExists;
/**
 * @deprecated Usage data reporting has been removed - returns undefined
 */
function readUsageDataSettings(_groupName) {
    return undefined;
}
exports.readUsageDataSettings = readUsageDataSettings;
